### ISSUE TITLE
Breach recovery issues

### ISSUE THREAD

**Issue by philipcmonk** [urbit/urbit#2952]
TL;DR if a new ship joined urbit-help (or any channel) before it finished with azimuth-tracker's initial update (about 10 minutes after boot), then it may have (50% chance) ended up in a broken state where it doesn't receive messages from urbit-help.  Workaround is to ;leave ~/~dopzod/urbit-help from chat-cli and then try to join again (possibly twice).

I'm writing this up because I want to make sure everyone understands how breaches are handled and their caveats.  Not that ignorance caused this issue — the two issues that came up are effectively bugs in my code.  But hopefully this will help people identify the sorts of problems that can arise during breach handling.

First, when a breach happens, Jael notifies its subscribers by turning over ~(tap in yen), which means they get notified in mug order (effectively random).  Consider what happens if Gall is notified before Ames.  Then an app hears a %kick, so it decides to resubscribe, so it sends a %watch via Ames, which enqueues that message and perhaps gives an effect to unix.  Then, before the event finishes, Ames hears about the breach, so it throws away all its state related to that ship — including the brand-new %watch.

Now, Gall thinks it's sent a %watch, so it has an unacknowledged outgoing subscription — but Ames knows nothing about it, and it'll never reach ~dopzod.  This is bad.  If Ames hears about the breach before Gall, then it works fine though.

You can resolve this by unsubscribing and resubscribing manually.  To solve it generally, I think the best solution is to send the notification first to Ames, then to Clay, then to Gall.  This way, gall apps can depend on the entire kernel knowing about the breach already.  Ames and Clay must be aware that later parts of the system don't know about the breach yet.

This is what happened with ~moltes-bitheb.  This is why it has many "breach peer" notifications early in its scrollback.  This suggests the second issue, which is that we shouldn't interpret the initial download of keys as evidence of a breach — we're just getting their keys for the first time, no need to throw anything away.  The code for deciding whether to notify a breach is small; you're welcome to identify the bug as an exercise:

```
      =/  maybe-point  (~(get by pos.zim) who)
      =/  =point  (fall maybe-point *point)
      ::  if changing rift upward, then signal a breach
      ::
      =?    ..feel
          ?&  ?=(%rift -.a-diff)
              (gth to.a-diff rift.point)
          ==
        %+  public-keys-give
          (subscribers-on-ship who)
        [%breach who]
      ...
```

I'm going to fix both of these issues and hopefully deploy them soon.  I'm holding urbit-os-v1.0.20 for now hoping that I can write this and test it sufficiently to push them both out.  In the meantime, I've resolved this on ~moltes with ;leave ~/~dopzod/urbit-help in the CLI and rejoining (possibly have to do it twice, or maybe I was impatient).

This is a possible cause for #2944

**Comment by joemfb** [comment]
@philipcmonk, this is obsolete?

**Comment by philipcmonk** [comment]
Yes, this was mitigated by #2953 and ultimately fixed in #4085
